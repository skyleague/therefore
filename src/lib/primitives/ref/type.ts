/* eslint-disable @typescript-eslint/no-unsafe-declaration-merging */
import type { NodeTrait } from '../../cst/mixin.js'
import { Node } from '../../cst/node.js'
import type { Intrinsic } from '../../cst/types.js'
import type { SchemaOptions } from '../base.js'

import { type ConstExpr, evaluate } from '@skyleague/axioms'

export type RefOptions = object

export class RefType<Reference extends Node = Node> extends Node {
    public override type = 'ref' as const
    public override children: [Node]

    public options: RefOptions = {}
    public declare infer: Reference['infer']
    public declare intrinsic: Intrinsic<Reference>

    public constructor(reference: ConstExpr<Reference>, options: SchemaOptions<RefOptions, Reference['infer']>) {
        super(options)
        this.options = options

        // we are lying to typescript here, but it's fine
        // there expression will be evaluated in the prepass, and errors hard if it's not
        this.children = [reference as Node]
    }

    public override hooks = {
        onLoad: [
            (self: Node): void => {
                self.children = self.children?.map(evaluate) as [Node]
            },
        ],
        onGenerate: [
            ({ children }: Node): void => {
                const [reference] = children ?? []
                if (reference !== undefined && reference.sourcePath === undefined) {
                    // this reference was not exported on its own
                    // so we just call it local if has no name
                    reference.name ??= 'local'
                }
            },
        ],
    }

    public static from<Reference extends Node>(
        ref: ConstExpr<Reference>,
        options: SchemaOptions<RefOptions>,
    ): RefType<Intrinsic<Reference>> {
        type IntrinsicType = Intrinsic<Reference>
        return new RefType<IntrinsicType>(ref as ConstExpr<IntrinsicType>, options)
    }
}

export interface RefType extends Node, NodeTrait {}
