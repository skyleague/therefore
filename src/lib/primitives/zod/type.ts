import type { ArrayType } from '../array/array.js'
import type { BooleanType } from '../boolean/boolean.js'
import type { ConstType } from '../const/const.js'
import type { IntersectionType } from '../intersection/intersection.js'
import type { NullableType } from '../nullable/nullable.js'
import type { NumberType } from '../number/number.js'
import type { ObjectType } from '../object/object.js'
import type { OptionalType } from '../optional/optional.js'
import type { RecordType } from '../record/record.js'
import type { StringType } from '../string/string.js'
import type { TupleType } from '../tuple/tuple.js'
import type { DiscriminatedUnionType, UnionType } from '../union/union.js'
import type { UnknownType } from '../unknown/unknown.js'

export type ZodSchema = { _def: { typeName: unknown }; _output: unknown }

export type ZodSchemaAsNode<T extends ZodSchema> = T['_def']['typeName'] extends 'ZodString'
    ? StringType
    : T['_def']['typeName'] extends 'ZodNumber'
      ? NumberType
      : T['_def']['typeName'] extends 'ZodBoolean'
        ? BooleanType
        : T['_def']['typeName'] extends 'ZodUndefined'
          ? ConstType<undefined>
          : T['_def']['typeName'] extends 'ZodNull'
            ? ConstType<null>
            : T['_def']['typeName'] extends 'ZodAny'
              ? UnknownType
              : T['_def']['typeName'] extends 'ZodUnknown'
                ? UnknownType
                : T extends { _def: { typeName: 'ZodArray' }; element: infer Element }
                  ? Element extends ZodSchema
                      ? ArrayType<ZodSchemaAsNode<Element>>
                      : never
                  : T extends { _def: { typeName: 'ZodObject' }; shape: infer Shape }
                    ? ObjectType<{
                          [K in keyof Shape]: Shape[K] extends ZodSchema ? ZodSchemaAsNode<Shape[K]> : never
                      }>
                    : T['_def']['typeName'] extends 'ZodUnion'
                      ? UnionType
                      : T['_def']['typeName'] extends 'ZodDiscriminatedUnion'
                        ? DiscriminatedUnionType
                        : T['_def']['typeName'] extends 'ZodIntersection'
                          ? IntersectionType
                          : T['_def']['typeName'] extends 'ZodTuple'
                            ? TupleType
                            : T extends { _def: { typeName: 'ZodRecord' }; element: infer Element }
                              ? RecordType<Element extends ZodSchema ? ZodSchemaAsNode<Element> : never>
                              : T extends { _def: { typeName: 'ZodLiteral' }; value: infer Value }
                                ? ConstType<Value>
                                : T extends { _def: { typeName: 'ZodOptional'; innerType: infer Schema } }
                                  ? Schema extends ZodSchema
                                      ? OptionalType<ZodSchemaAsNode<Schema>>
                                      : never
                                  : T extends { _def: { typeName: 'ZodNullable'; innerType: infer Schema } }
                                    ? Schema extends ZodSchema
                                        ? NullableType<ZodSchemaAsNode<Schema>>
                                        : never
                                    : T['_def']['typeName'] extends 'ZodDefault'
                                      ? BooleanType //
                                      : T extends {
                                              _def: { typeName: 'ZodEffects' }
                                              innerType: (...args: unknown[]) => infer Schema
                                          }
                                        ? Schema extends ZodSchema
                                            ? ZodSchemaAsNode<Schema>
                                            : never
                                        : never
