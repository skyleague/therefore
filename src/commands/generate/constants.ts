import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
export const generatedAjv = `Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code`
