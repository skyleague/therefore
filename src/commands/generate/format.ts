import path from 'node:path'
import { Biome, Distribution } from '@biomejs/js-api'
import type { ThereforeOutputType } from '../../lib/output/types.js'

// biome-ignore lint/suspicious/noExplicitAny: it's a third-party library
export type Prettier = any
export async function maybeLoadPrettier() {
    try {
        // @ts-ignore
        return await import('prettier')
    } catch {
        return undefined
    }
}

const biome = await (async () => {
    const local = await Biome.create({
        distribution: Distribution.NODE,
    })
    try {
        const baseConfig = (await import(`file:///${process.cwd().replace(/\\/g, '/')}/biome.json`)).default
        if ('extends' in baseConfig && Array.isArray(baseConfig.extends)) {
            for (const extend of baseConfig.extends) {
                const extendedConfig = (await import(extend)).default
                local.applyConfiguration(extendedConfig)
            }
        }
        local.applyConfiguration(baseConfig)
    } catch (_e) {
        console.debug('No biome.json found')
    }

    return local
})()

export async function formatFile({
    prettier,
    input,
    file,
    type,
}: {
    prettier: Prettier
    input: string
    file: string
    type: ThereforeOutputType
}): Promise<string> {
    if (prettier !== undefined) {
        const options = await prettier.resolveConfig(file)
        return prettier.format(input, {
            ...(options ?? {}),
            parser: type,
        })
    }
    const formatted = biome.formatContent(input, {
        // this virual path might ignore the file alltogether
        // filePath: file,
        filePath: `src/${path.dirname(file)}.${file.split('.').pop()}`,
    })
    return formatted.content
}
