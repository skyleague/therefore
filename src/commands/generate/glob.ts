import { Nothing, fstat, isJust, isRight, normalizePath } from '@skyleague/axioms'
import fastGlob from 'fast-glob'

import path from 'node:path'

// should give the same glob behaviour as Prettier
// https://github.com/prettier/prettier/blob/main/src/cli/expand-patterns.js
// also heavily inspired by the logic there :)
export async function expandGlobs({
    patterns,
    ignore = [],
    cwd,
    extension,
}: {
    patterns: string[]
    ignore?: string[]
    cwd: string
    extension: string
}): Promise<string[]> {
    const ignoredDirectories = ['.git', '.svn', '.hg', 'node_modules']
    const globOptions = {
        dot: true,
        ignore: [...ignoredDirectories.map((dir) => `**/${dir}`), ...ignore],
    }

    const entries: string[] = (
        await Promise.allSettled(
            patterns.map(async (pattern) => {
                const absolutePath = path.resolve(cwd, pattern)
                if (ignoredDirectories.some((i) => pattern.includes(i))) {
                    return Nothing
                }

                const eitherStat = await fstat(absolutePath)
                if (isRight(eitherStat) && isJust(eitherStat.right)) {
                    if (eitherStat.right.isFile()) {
                        return fastGlob.escapePath(normalizePath(pattern))
                    }
                    if (eitherStat.right.isDirectory()) {
                        return `${fastGlob.escapePath(normalizePath(path.relative(cwd, absolutePath)))}/**/*${extension}`
                    }

                    return Nothing
                }
                if (pattern.startsWith('!')) {
                    globOptions.ignore.push(normalizePath(pattern.slice(1)))
                    return Nothing
                }
                return normalizePath(pattern)
            }),
        )
    )
        .map((result) => (result.status === 'fulfilled' ? result.value : Nothing))
        .filter(isJust)
    return await fastGlob(entries, globOptions)
}
