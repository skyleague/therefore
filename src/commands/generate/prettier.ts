import type { ThereforeOutputType } from './types.js'

import { Nothing, isNothing } from '@skyleague/axioms'

export async function maybeLoadPrettier() {
    try {
        return await import('prettier')
    } catch {
        return Nothing
    }
}

export async function formatFile(
    prettier: Awaited<ReturnType<typeof maybeLoadPrettier>>,
    input: string,
    file: string,
    type: ThereforeOutputType
): Promise<string> {
    if (isNothing(prettier)) {
        return input
    }
    const options = await prettier.resolveConfig(file)
    return prettier.format(input, {
        ...(options ?? {}),
        ...(type === 'jsonschema' ? { parser: 'json' } : {}),
    })
}
