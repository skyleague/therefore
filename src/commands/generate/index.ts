import { generate } from './generate.js'

import { constants } from '../../lib/constants.js'

import type { Argv } from 'yargs'

export function builder(yargs: Argv) {
    return yargs
        .option('files', {
            alias: 'f',
            demandOption: true,
            describe: 'globs to scan for schemas',
            type: 'array',
        })
        .option('ignore-pattern', {
            alias: 'i',
            demandOption: true,
            default: ['**/*.d.ts', 'node_modules'],
            describe: 'globs to exclude',
            type: 'array',
        })
        .option('clean', {
            default: false,
            type: 'boolean',
        })
}

export async function handler(argv: ReturnType<typeof builder>['argv']): Promise<void> {
    const { files = [], 'ignore-pattern': ignorePatterns, clean } = await argv

    await generate({
        globs: files.map((f) => f.toString()),
        ignore: ignorePatterns.map((p) => p.toString()),
        extension: constants.schemaExtension,
        clean,
    })
}

export default {
    command: ['generate', '*'],
    describe: 'generate therefore types',
    builder,
    handler,
}
