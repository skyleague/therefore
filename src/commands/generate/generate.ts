import { generatedAjv, generatedBy } from './constants.js'
import { expandGlobs } from './glob.js'
import { formatFile, maybeLoadPrettier } from './prettier.js'
import { resolveTypescriptSchema } from './resolver.js'
import type { FileDefinition, OutputFile, ReferenceData, ThereforeOutputType } from './types.js'

import { getExtension, replaceExtension } from '../../common/template/path.js'
import { renderTemplate } from '../../common/template/template.js'
import type { ThereforeNode } from '../../lib/cst/cst.js'
import type { ThereforeCst } from '../../lib/primitives/types.js'
import { isThereforeExport } from '../../lib/primitives/types.js'
import { toJsonSchema } from '../../lib/visitor/jsonschema/jsonschema.js'
import { prepass } from '../../lib/visitor/prepass/prepass.js'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript.js'

import { entriesOf, enumerate, groupBy, isDefined, range, second } from '@skyleague/axioms'
import decamelize from 'decamelize'

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, ThereforeNode | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace(/\.ts$/, '.js')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, ThereforeNode | unknown>
}

export async function loadSymbol({
    srcPath,
    basePath,
    symbol,
    sourceSymbol,
    compile,
    definitions,
    entry,
    fileSuffix,
    filePath,
    outputFileRename,
    appendSourceSymbolName: appendSourceName = false,
    rootPath = srcPath,
}: {
    srcPath: string
    basePath: string
    symbol: ThereforeCst
    sourceSymbol: string
    compile: boolean
    definitions: Record<string, FileDefinition | undefined>
    entry: string
    fileSuffix?: string | undefined
    filePath?: string | undefined
    outputFileRename: (path: string) => string
    appendSourceSymbolName?: boolean
    rootPath?: string
}) {
    const simplified = prepass(symbol)
    let targetDir = '.'
    if (simplified.type === 'custom') {
        if (simplified.value.fileSuffix !== undefined) {
            fileSuffix = simplified.value.fileSuffix
        }
        if (simplified.value.filePath !== undefined) {
            const dir = path.dirname(simplified.value.filePath)
            const basename = path.basename(simplified.value.filePath)
            filePath = basename
            targetDir = dir
        }
    }

    const { base: baseName, dir: baseDir } = path.parse(srcPath)
    const ext = getExtension(srcPath) ?? ''
    const targetBaseDir = path.relative(basePath, path.join(baseDir, targetDir))
    const targetSrcPath = `${targetBaseDir}/${(filePath ?? baseName).replace(ext, '')}${ext}`
    const targetPath = fileSuffix !== undefined ? replaceExtension(targetSrcPath, fileSuffix) : outputFileRename(targetSrcPath)

    definitions[targetPath] ??= {
        srcPath: path.relative(basePath, rootPath),
        targetPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }

    const file = definitions[targetPath]!
    // check if symbol is not seen yet
    if (file.symbols.find((s) => s.uuid === simplified.uuid)) {
        return
    }
    const symbolName = simplified.name ?? sourceSymbol

    sourceSymbol = appendSourceName && simplified.name !== undefined ? `${sourceSymbol}.${simplified.name}` : symbolName
    const schemaName = decamelize(symbolName, { separator: '-' })
    const schemaFile = `./schemas/${schemaName}.schema.json`
    const compiledFile = `./schemas/${schemaName}.schema.js`

    const nodeIsCompiled = simplified.description.validator?.compile ?? compile

    const jsonschema = await toJsonSchema(simplified, nodeIsCompiled)
    const { definition, subtrees } = toTypescriptDefinition({ sourceSymbol, symbolName, schema: simplified })

    file.symbols.push(
        ...Object.values(definition.locals ?? {})
            .filter(isDefined)
            .map((local) => ({
                uuid: local.uuid,
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
    )

    file.symbols.push({
        uuid: simplified.uuid,
        symbolName,
        definition: definition,
        schemaFile,
        compiledFile: nodeIsCompiled ? compiledFile : undefined,
        typeOnly: simplified.description.validator?.enabled !== true,
    })

    if (definition.isExported && simplified.description.validator?.enabled) {
        const cleanSchemasFolder = () => {
            const targetFolder = path.dirname(path.join(targetBaseDir, compiledFile))
            if (fs.existsSync(targetFolder)) {
                console.warn(`Cleaning ${targetFolder}`)
                fs.rmSync(targetFolder, { force: true, recursive: true })
            }
        }
        if (jsonschema.compiled) {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, compiledFile),
                content: `/**\n * ${generatedAjv} \n * eslint-disable\n */\n${jsonschema.code}`,
                prettify: false,
                type: 'validator',
                clean: cleanSchemasFolder,
            })
        } else {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, schemaFile),
                content: JSON.stringify(jsonschema.schema, null, 2),
                prettify: true,
                type: 'jsonschema',
                clean: cleanSchemasFolder,
            })
        }

        console.debug(` - found ${definition.symbolName}`)
    }
    for (const { node: subSymbol, fileSuffix: subFileSuffix, filePath: subFilePath } of subtrees) {
        await loadSymbol({
            symbol: subSymbol,
            basePath,
            sourceSymbol,
            compile,
            definitions,
            srcPath: path.join(basePath, targetPath),
            rootPath: srcPath,
            fileSuffix: subFileSuffix,
            filePath: subFilePath,
            entry,
            outputFileRename,
            appendSourceSymbolName: true,
        })
    }
}

export async function scanModule({
    entry,
    srcPath,
    basePath,
    compile,
    definitions,
    require = requireModule,
    outputFileRename,
}: {
    entry: string
    srcPath: string
    basePath: string
    compile: boolean
    definitions: Record<string, FileDefinition>
    require?: (module: string) => Promise<Record<string, ThereforeNode | unknown>>
    outputFileRename: (path: string) => string
}) {
    const module = await require(entry)

    for (const [sourceSymbol, symbolPromise] of Object.entries(module)) {
        const symbol = await symbolPromise

        if (!isThereforeExport(symbol)) {
            continue
        }

        await loadSymbol({ symbol, sourceSymbol, compile, definitions, srcPath, entry, basePath, outputFileRename })
    }
}

export async function scanFiles({
    files,
    basePath,
    compile,
    outputFileRename,
}: {
    files: string[]
    basePath: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<Record<string, FileDefinition>> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const srcPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = srcPath.substr(0, srcPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            await scanModule({ entry, basePath, srcPath, compile, definitions, outputFileRename })
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

export async function compileOutputFiles(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        targetPath: string
        template: string
        data: Record<string, ReferenceData>
        type: ThereforeOutputType
        prettify: boolean
        clean?: () => void
    }[]
> {
    const outputFiles: OutputFile[] = []
    const definitions = await scanFiles({ files: entries, basePath: cwd, compile, outputFileRename })

    const localReferences = {}

    for (const targetPath of Object.keys(definitions)) {
        resolveTypescriptSchema({ definitions, targetPath, cwd, outputFiles, localReferences })
    }

    return compile ? outputFiles : outputFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const outputFiles = await compileOutputFiles(entries, {
        cwd,
        outputFileRename,
        compile,
    })

    if (clean) {
        for (const of of outputFiles) {
            if (of.clean !== undefined) {
                of.clean()
            }
        }
    }

    const prettier = maybeLoadPrettier()

    const data = outputFiles.reduce<Record<string, ReferenceData>>(
        (agg, x) => ({
            ...agg,
            ...x.data,
        }),
        {}
    )

    const duplicates = entriesOf(groupBy(entriesOf(data), ([, values]) => `${values.srcPath}${values.symbolName}`)).filter(
        ([, records]) => records.length > 1
    )
    for (const duplicateGroup of duplicates.map((d) => second(d))) {
        for (const [i, [key]] of enumerate(duplicateGroup)) {
            // data[key].referenceName = `${data[key].referenceName}${i}`
            data[key]!.symbolName = `${data[key]!.symbolName}${i}`
            data[key]!.uniqueSymbolName = `${data[key]!.uniqueSymbolName}${i}`
        }
    }

    const variablesMap = Object.fromEntries(
        Object.entries(data)
            .map(([key, vals]) => Object.entries(vals).map(([varName, value]) => [`${key}:${varName}`, value] as const))
            .flat()
    )

    for (const { targetPath, template, type, prettify } of outputFiles) {
        fs.mkdirSync(path.dirname(targetPath), { recursive: true })
        // allow nested references up to 3 deep
        let contents = template
        for (const _ of range(3)) {
            contents = renderTemplate(contents, variablesMap)
        }
        fs.writeFileSync(targetPath, prettify ? await formatFile(prettier, contents, targetPath, type) : contents)
    }
}
