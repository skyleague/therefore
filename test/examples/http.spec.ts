import { Headers, Query } from '../../examples/http/http.type.js'
import { compileOutput } from '../../src/commands/generate/generate.js'

import { forAll, integer, isNumber, record, string, tuple, unknown } from '@skyleague/axioms'
import { expect, it } from 'vitest'

it('http', async () => {
    expect(
        await compileOutput(['examples/http/http.schema.ts'], {
            cwd: process.cwd(),
        }),
    ).toMatchSnapshot()
})

it('headers', () => {
    forAll(tuple(record(unknown()), string()), ([d, x]) => {
        const h = { ...d, authorization: x }
        return Headers.is(h)
    })
})

it('query', () => {
    forAll(tuple(record(unknown()), integer()), ([d, x]) => {
        const q = { ...d, limit: x.toString() }
        return Query.is(q) && isNumber(q.limit)
    })
})

it('schema', () => {
    expect(Query.schema).toMatchInlineSnapshot(`
      {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": true,
        "properties": {
          "limit": {
            "type": "number",
          },
        },
        "required": [
          "limit",
        ],
        "title": "Query",
        "type": "object",
      }
    `)
})
