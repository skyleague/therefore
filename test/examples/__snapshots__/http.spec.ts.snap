// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`http 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/http/http.type.ts": "/**
 * Generated by @skyleague/therefore@v1.0.0-local
 * Do not manually touch this
 */
/* eslint-disable */

import type { DefinedError, ValidateFunction } from 'ajv'

import { validate as DepValidator } from './schemas/dep.schema.js'
import { validate as HeadersValidator } from './schemas/headers.schema.js'
import { validate as QueryValidator } from './schemas/query.schema.js'

export interface Dep {
    authorization: string
}

export const Dep = {
    validate: DepValidator as ValidateFunction<Dep>,
    get schema() {
        return Dep.validate.schema
    },
    get errors() {
        return Dep.validate.errors ?? undefined
    },
    is: (o: unknown): o is Dep => Dep.validate(o) === true,
    parse: (o: unknown): { right: Dep } | { left: DefinedError[] } => {
        if (Dep.is(o)) {
            return { right: o }
        }
        return { left: (Dep.errors ?? []) as DefinedError[] }
    },
} as const

export interface Headers {
    authorization: string
}

export const Headers = {
    validate: HeadersValidator as ValidateFunction<Headers>,
    get schema() {
        return Headers.validate.schema
    },
    get errors() {
        return Headers.validate.errors ?? undefined
    },
    is: (o: unknown): o is Headers => Headers.validate(o) === true,
    parse: (o: unknown): { right: Headers } | { left: DefinedError[] } => {
        if (Headers.is(o)) {
            return { right: o }
        }
        return { left: (Headers.errors ?? []) as DefinedError[] }
    },
} as const

export interface Query {
    limit: number
}

export const Query = {
    validate: QueryValidator as ValidateFunction<Query>,
    get schema() {
        return Query.validate.schema
    },
    get errors() {
        return Query.validate.errors ?? undefined
    },
    is: (o: unknown): o is Query => Query.validate(o) === true,
    parse: (o: unknown): { right: Query } | { left: DefinedError[] } => {
        if (Query.is(o)) {
            return { right: o }
        }
        return { left: (Query.errors ?? []) as DefinedError[] }
    },
} as const
",
    "examples/http/schemas/dep.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"Dep","type":"object","properties":{"authorization":{"type":"string"}},"required":["authorization"],"additionalProperties":true};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){let missing0;if((data.authorization === undefined) && (missing0 = "authorization")){validate10.errors = [{instancePath,schemaPath:"#/required",keyword:"required",params:{missingProperty: missing0},message:"must have required property '"+missing0+"'"}];return false;}else {if(data.authorization !== undefined){let data0 = data.authorization;if(typeof data0 !== "string"){let dataType0 = typeof data0;let coerced0 = undefined;if(!(coerced0 !== undefined)){if(dataType0 == "number" || dataType0 == "boolean"){coerced0 = "" + data0;}else if(data0 === null){coerced0 = "";}else {validate10.errors = [{instancePath:instancePath+"/authorization",schemaPath:"#/properties/authorization/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}}if(coerced0 !== undefined){data0 = coerced0;if(data !== undefined){data["authorization"] = coerced0;}}}}}}else {validate10.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
    "examples/http/schemas/headers.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"Headers","type":"object","properties":{"authorization":{"type":"string"}},"required":["authorization"],"additionalProperties":true};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){let missing0;if((data.authorization === undefined) && (missing0 = "authorization")){validate10.errors = [{instancePath,schemaPath:"#/required",keyword:"required",params:{missingProperty: missing0},message:"must have required property '"+missing0+"'"}];return false;}else {if(data.authorization !== undefined){let data0 = data.authorization;if(typeof data0 !== "string"){let dataType0 = typeof data0;let coerced0 = undefined;if(!(coerced0 !== undefined)){if(dataType0 == "number" || dataType0 == "boolean"){coerced0 = "" + data0;}else if(data0 === null){coerced0 = "";}else {validate10.errors = [{instancePath:instancePath+"/authorization",schemaPath:"#/properties/authorization/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}}if(coerced0 !== undefined){data0 = coerced0;if(data !== undefined){data["authorization"] = coerced0;}}}}}}else {validate10.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
    "examples/http/schemas/query.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"Query","type":"object","properties":{"limit":{"type":"number"}},"required":["limit"],"additionalProperties":true};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){let missing0;if((data.limit === undefined) && (missing0 = "limit")){validate10.errors = [{instancePath,schemaPath:"#/required",keyword:"required",params:{missingProperty: missing0},message:"must have required property '"+missing0+"'"}];return false;}else {if(data.limit !== undefined){let data0 = data.limit;if(!((typeof data0 == "number") && (isFinite(data0)))){let dataType0 = typeof data0;let coerced0 = undefined;if(!(coerced0 !== undefined)){if(dataType0 == "boolean" || data0 === null
              || (dataType0 == "string" && data0 && data0 == +data0)){coerced0 = +data0;}else {validate10.errors = [{instancePath:instancePath+"/limit",schemaPath:"#/properties/limit/type",keyword:"type",params:{type: "number"},message:"must be number"}];return false;}}if(coerced0 !== undefined){data0 = coerced0;if(data !== undefined){data["limit"] = coerced0;}}}}}}else {validate10.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
  },
}
`;
